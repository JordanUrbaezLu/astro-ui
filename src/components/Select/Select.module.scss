@use "../../utility/globals";

.selectContainer {
  font-family: globals.$font-family;
  position: relative;
  user-select: none;
  width: 142px;
}

.select {
  align-items: center;
  border: 3px solid globals.$primary-background-color;
  border-radius: 5px;
  column-gap: 6px;
  display: flex;
  font-size: 16px;
  outline: none !important;
  padding: 10px 6px;
}

.selectInput {
  align-items: center;
  border: 2px solid globals.$primary-background-color;
  border-radius: 6px;
  box-sizing: border-box;
  display: flex;
  font-size: 18px;
  height: 46px;
  padding: 10px;
  transition: all 130ms ease-in;
  width: 142px;

  &.isOpen {
    box-shadow: 0 0 5px 5px globals.$primary-box-shadow;
  }
}

.label {
  color: globals.$primary-background-color;
  font-family: globals.$font-family;
  font-size: 14px;
  padding-left: 8px;
  user-select: none;
}

.selectIcon {
  background-color: transparent !important;
  color: globals.$primary-background-color !important;
  left: 100px;
  position: absolute;
  top: 4px;
  z-index: 1;

  &:hover,
  &:focus {
    background-color: globals.$primary-text-color !important;
  }

  &:active {
    background-color: globals.$primary-background-color-active !important;
    color: globals.$primary-text-color !important;
  }
}

.enter {
  opacity: 0;
  transform: translateY(-6px);
}

.enterActive {
  opacity: 1;
  transform: translateY(0);
  transition: all 200ms ease-in-out;
}

.exit {
  opacity: 1;
  transform: translateY(0);
}

.exitActive {
  opacity: 0;
  transform: translateY(-6px);
  transition: all 200ms ease-in-out;
}
