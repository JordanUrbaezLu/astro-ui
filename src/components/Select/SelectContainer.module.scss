@use "../../utility/globals";

.selectOptionsMenu {
  background-color: globals.$primary-background-color;
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  margin-top: 2px;
  padding: 4px 0;
  position: absolute;
  width: 142px;
  z-index: 1;
}
