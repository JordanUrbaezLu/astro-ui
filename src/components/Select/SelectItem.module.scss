@use "../../utility/globals";

.selectItem {
  color: white;
  cursor: pointer;
  font-family: globals.$font-family;
  font-size: 16px;
  outline: none;
  padding: 5px 5px 5px 12px;
  text-align: left;

  &.selected {
    background-color: rgb(97 70 109);

    &:hover,
    &:focus {
      background-color: rgb(110 97 115);
    }

    &:active {
      background-color: globals.$primary-background-color-active;
    }
  }

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
  }
}
