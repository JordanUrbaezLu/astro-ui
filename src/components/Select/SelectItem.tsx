import * as React from "react";
interface SelectItemProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The content for the SelectItem
   */
  children: string;
}

/**
 * @public
 */
export const SelectItem: React.FC<SelectItemProps> = (props) => {
  const { className, children, ...rest } = props;
  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

SelectItem.displayName = "SelectItem";
