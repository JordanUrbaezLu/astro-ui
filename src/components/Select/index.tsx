/* istanbul ignore file */
export * from "./Select";
export * from "./SelectItem";
