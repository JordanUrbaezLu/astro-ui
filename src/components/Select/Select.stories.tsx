import { Story } from "@storybook/react";
import * as React from "react";
import { Select, SelectProps } from "./Select";
import { SelectItem } from "./SelectItem";
import { within, userEvent } from "@storybook/testing-library";
import { allowLayoutCalculations } from "../../../.storybook/utilities/allowLayoutCalculations";

export default {
  component: Select,
  parameters: {
    chromatic: { delay: 2000 },
  },
  title: "Components/Inputs/Select",
};

const Template: Story<SelectProps> = (args) => {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  return (
    <Select
      {...args}
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
    >
      <SelectItem>CA</SelectItem>
      <SelectItem>MA</SelectItem>
      <SelectItem>FL</SelectItem>
    </Select>
  );
};

export const Default = Template.bind({});

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: "State",
};
Placeholder.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};

export const Selected = Template.bind({});
Selected.args = {
  selected: 1,
};
Selected.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};

export const WithManySelectItems: Story<SelectProps> = (args) => {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  return (
    <Select
      {...args}
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
    >
      <SelectItem>10</SelectItem>
      <SelectItem>20</SelectItem>
      <SelectItem>30</SelectItem>
      <SelectItem>40</SelectItem>
      <SelectItem>50</SelectItem>
      <SelectItem>60</SelectItem>
      <SelectItem>70</SelectItem>
      <SelectItem>80</SelectItem>
      <SelectItem>90</SelectItem>
    </Select>
  );
};

WithManySelectItems.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};
