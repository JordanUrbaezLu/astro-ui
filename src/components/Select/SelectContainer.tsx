import classNames from "classnames";
import * as React from "react";
import { SelectProps } from "./Select";
import styles from "./SelectContainer.module.scss";
import itemStyles from "./SelectItem.module.scss";
import { useOnClickOutside, useOnKeyDown } from "../../hooks";
import { SelectItem } from "./SelectItem";
import { useSelectFocus } from "../../hooks/useSelectFocus";

export interface SelectContainerProps extends SelectProps {
  /**
   * The callback fired when requesting to change selected index
   */
  onSetSelectedIndex: (index: number) => void;
  /**
   * The ref for the Select
   */
  selectRef: React.RefObject<HTMLDivElement>;
  /**
   * The trigger ref for the Select
   */
  triggerRef: React.RefObject<HTMLButtonElement>;
}

/**
 * @private
 */
export const SelectContainer: React.FC<SelectContainerProps> = (
  props
) => {
  const {
    children,
    selected,
    onClose,
    onSetSelectedIndex,
    selectRef,
    triggerRef,
  } = props;

  const selectContainerRef = React.useRef<HTMLDivElement>(null);

  const { focusNextItem, focusPreviousItem, focusedSelectItemIndex } =
    useSelectFocus(children, selectContainerRef, selected);

  useOnClickOutside(selectRef, onClose);
  useOnKeyDown(["Escape"], () => {
    onClose();
    triggerRef.current?.focus();
  });

  return (
    <div
      className={styles.selectOptionsMenu}
      ref={selectContainerRef}
      role="listbox"
    >
      {React.Children.map(children, (child, index) => {
        const selectItem = child as React.ReactElement;

        if (selectItem.type === SelectItem) {
          return React.cloneElement(selectItem, {
            className: classNames(
              itemStyles.selectItem,
              selected === index && itemStyles.selected
            ),
            key: index,
            onClick: () => {
              onClose();
              onSetSelectedIndex(index);
            },
            onKeyDown: (event: React.KeyboardEvent<HTMLElement>) => {
              if (event.code === "Space") {
                onClose();
                onSetSelectedIndex(index);
                triggerRef.current?.focus();
              }
              if (event.code === "ArrowUp") {
                focusPreviousItem();
              }
              if (event.code === "ArrowDown") {
                focusNextItem();
              }
            },
            role: "listitem",
            tabIndex:
              index === focusedSelectItemIndex ? 0 : undefined,
          });
        } else {
          return null;
        }
      })}
    </div>
  );
};

SelectContainer.displayName = "SelectContainer";
