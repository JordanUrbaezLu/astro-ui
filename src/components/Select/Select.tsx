import { CSSTransition } from "react-transition-group";
import styles from "./Select.module.scss";
import { IoMdArrowDropup, IoMdArrowDropdown } from "react-icons/io";
import * as React from "react";
import { SelectContainer } from "./SelectContainer";
import classNames from "classnames";
import { IconButton } from "../IconButton";

export interface SelectProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The content for the Select
   */
  children: React.ReactNode;
  /**
   * If the Select is open
   *
   * @default false
   */
  isOpen?: boolean;
  /**
   * The callback fired when the Select closes
   */
  onClose: () => void;
  /**
   * The callback fired when the Select opens
   */
  onOpen: () => void;
  /**
   * The placeholder for the Select
   */
  placeholder?: string;
  /**
   * The initial selected item's index for the Select
   */
  selected?: number;
}

/**
 * @public
 *
 * @description
 *
 * The Select component displays a dropdown of selectable options.
 */
export const Select: React.FC<SelectProps> = (props) => {
  const {
    className,
    children,
    isOpen = false,
    onClose,
    onOpen,
    placeholder,
    selected,
    ...rest
  } = props;

  const [selectedIndex, setSelectedIndex] =
    React.useState<number | undefined>(selected);
  const [value, setValue] = React.useState<string>();
  const selectRef = React.useRef<HTMLDivElement>(null);
  const triggerRef = React.useRef<HTMLButtonElement>(null);

  React.useEffect(() => {
    if (selectedIndex !== undefined) {
      setValue(
        React.Children.toArray(children)[selectedIndex] as string
      );
    }
  }, [selectedIndex]);

  const onSetSelectedIndex = (index: number) => {
    setSelectedIndex(index);
  };

  return (
    <div
      className={classNames(className, styles.selectContainer)}
      ref={selectRef}
      {...rest}
    >
      <div
        className={classNames(
          styles.selectInput,
          isOpen && styles.isOpen
        )}
        onClick={() => {
          if (isOpen) {
            onClose();
          } else {
            onOpen();
          }
        }}
      >
        {value ?? <div style={{ color: "gray" }}>{placeholder}</div>}
      </div>
      {isOpen ? (
        <IconButton
          ariaLabel="Chevron Up"
          className={styles.selectIcon}
          content={<IoMdArrowDropup />}
          onClick={onClose}
          ref={triggerRef}
        />
      ) : (
        <IconButton
          ariaLabel="Chevron Down"
          className={styles.selectIcon}
          content={<IoMdArrowDropdown />}
          onClick={onOpen}
          ref={triggerRef}
        />
      )}
      <CSSTransition
        in={isOpen}
        timeout={200}
        mountOnEnter
        unmountOnExit
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
      >
        <SelectContainer
          selected={selectedIndex}
          isOpen={isOpen}
          onOpen={onOpen}
          onClose={onClose}
          onSetSelectedIndex={onSetSelectedIndex}
          selectRef={selectRef}
          triggerRef={triggerRef}
        >
          {children}
        </SelectContainer>
      </CSSTransition>
    </div>
  );
};

Select.displayName = "Select";
