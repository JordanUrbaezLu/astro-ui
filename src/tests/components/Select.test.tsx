import {
  fireEvent,
  render,
  screen,
  createEvent,
  act,
} from "@testing-library/react";
import { Select } from "../../components/Select/Select";
import { SelectItem } from "../../components/Select/SelectItem";
import { axe } from "jest-axe";

describe("Select", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Select onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Select onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
      </Select>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should open SelectContainer on icon click.", () => {
    render(
      <Select
        isOpen
        onClose={jest.fn()}
        onOpen={jest.fn()}
        placeholder="Number"
      >
        <SelectItem>30</SelectItem>
      </Select>
    );

    act(() => screen.getByRole("button").click());

    expect(screen.getByText("30")).toBeInTheDocument();

    act(() => screen.getByRole("listitem").click());
  });

  test("Should open SelectContainer on placeholder click.", () => {
    render(
      <Select
        onClose={jest.fn()}
        onOpen={jest.fn()}
        placeholder="Placeholder"
      >
        <SelectItem>30</SelectItem>
      </Select>
    );

    act(() => screen.getByText("Placeholder").click());
  });

  test("Should close SelectConatiner on label click.", () => {
    render(
      <Select
        isOpen
        onClose={jest.fn()}
        onOpen={jest.fn()}
        selected={0}
      >
        <SelectItem>30</SelectItem>
      </Select>
    );

    screen.getAllByText("30")[0].click();
  });

  test("Should close Select on click outside.", () => {
    render(
      <Select onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
      </Select>
    );

    fireEvent.mouseDown(document.body);
  });

  test("Should trigger onClose on Escape key down.", () => {
    const onClose = jest.fn();

    render(
      <Select isOpen onOpen={jest.fn} onClose={onClose}>
        <SelectItem>30</SelectItem>
      </Select>
    );

    fireEvent.keyDown(document.body, { key: "Escape" });

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should trigger key down events correctly.", () => {
    const onClose = jest.fn();

    render(
      <Select isOpen onClose={onClose} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
        <SelectItem>40</SelectItem>
      </Select>
    );

    const selectItem = screen.getAllByRole("listitem")[0];

    const spaceEvent = createEvent.keyDown(selectItem, {
      code: "Space",
    });

    fireEvent(selectItem, spaceEvent);

    expect(onClose).toHaveBeenCalledTimes(1);

    const arrowUpEvent = createEvent.keyDown(selectItem, {
      code: "ArrowUp",
    });

    fireEvent(selectItem, arrowUpEvent);

    expect(onClose).toHaveBeenCalledTimes(1);

    const arrowDownEvent = createEvent.keyDown(selectItem, {
      code: "ArrowDown",
    });

    fireEvent(selectItem, arrowDownEvent);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should focus select items correctly.", () => {
    render(
      <Select isOpen onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
        <SelectItem>40</SelectItem>
      </Select>
    );

    const firstItem = screen.getAllByRole("listitem")[0];

    const secondItem = screen.getAllByRole("listitem")[1];

    expect(firstItem).toHaveFocus();

    const event1 = createEvent.keyDown(firstItem, {
      code: "ArrowDown",
    });

    fireEvent(firstItem, event1);

    expect(secondItem).toHaveFocus();

    const event2 = createEvent.keyDown(secondItem, {
      code: "ArrowDown",
    });

    fireEvent(secondItem, event2);

    expect(firstItem).toHaveFocus();

    const event3 = createEvent.keyDown(firstItem, {
      code: "ArrowUp",
    });

    fireEvent(firstItem, event3);

    expect(secondItem).toHaveFocus();

    const event4 = createEvent.keyDown(secondItem, {
      code: "ArrowUp",
    });

    fireEvent(secondItem, event4);

    expect(firstItem).toHaveFocus();
  });
});

describe("Props", () => {
  test("Should render children correctly.", () => {
    render(
      <Select
        isOpen
        onClose={jest.fn()}
        onOpen={jest.fn()}
        placeholder="Select"
      >
        <SelectItem>30</SelectItem>
        <div>40</div>
      </Select>
    );
    expect(screen.getByRole("listbox")).not.toHaveTextContent("40");
  });
  test("Should render label correctly.", () => {
    render(
      <Select
        onClose={jest.fn()}
        onOpen={jest.fn()}
        placeholder="Select"
      >
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(screen.getByText("Select")).toBeInTheDocument();
  });

  test("Should render correct Chevron when open.", () => {
    render(
      <Select isOpen onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(screen.getByLabelText("Chevron Up")).toBeInTheDocument();
  });

  test("Should render correct Chevron when closed.", () => {
    render(
      <Select onClose={jest.fn()} onOpen={jest.fn()}>
        <SelectItem>30</SelectItem>
      </Select>
    );
    expect(screen.getByLabelText("Chevron Down")).toBeInTheDocument();
  });
});
