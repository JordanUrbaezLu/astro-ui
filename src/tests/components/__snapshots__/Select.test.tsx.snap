// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should render component correctly. 1`] = `
<div
  class="selectContainer"
>
  <div
    class="selectInput"
  >
    <div
      style="color: gray;"
    />
  </div>
  <button
    aria-label="Chevron Down"
    class="selectIcon iconButton medium primary"
  >
    <div
      class="container"
      role="presentation"
    />
    <svg
      fill="currentColor"
      height="22"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="22"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M128 192l128 128 128-128z"
      />
    </svg>
  </button>
</div>
`;
